client/components/data/query-sites/index.jsx

```diff
@@ -3,19 +3,31 @@
  */
 import { Component, PropTypes } from 'react';
 import { connect } from 'react-redux';
-import { bindActionCreators } from 'redux';
 
 /**
  * Internal dependencies
  */
-import { isRequestingSites } from 'state/sites/selectors';
-import { requestSites } from 'state/sites/actions';
+import { isRequestingSites, isRequestingSite } from 'state/sites/selectors';
+import { requestSites, requestSite } from 'state/sites/actions';
 
 class QuerySites extends Component {
-
 	componentWillMount() {
-		if ( ! this.props.requestingSites ) {
-			this.props.requestSites();
+		this.request( this.props );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.siteId !== this.props.siteId ) {
+			this.request( nextProps );
+		}
+	}
+
+	request( props ) {
+		if ( ! props.siteId && ! props.requestingSites ) {
```

Comment:
>Do we need to be careful about the case where we may need to wait until a siteId is available? We might accidentally trigger an all sites request if we pass through a falsy value before we pass through a siteId.

Yes, but I think that should be on the developer to ensure that if they're intending to request a single site that they not render the `<QuerySites />` component until the site ID is ready. That said, we could/should probably have better documentation about this behavior.